INVALID_ARGUMENT: Digest mismatch

Having configured Bazel to use `SHA256` 
```
startup --host_jvm_args=-Dbazel.DigestFunction=SHA256
```
and build farm server to use SHA256 by adding
```
hash_function = SHA256
```

I'm getting the following exception when trying to build using remote build:
```
ERROR: /private/var/tmp/_bazel_ttsugrii/98bcb5c7cfb9e117ec3600033a1548b4/external/com_google_protobuf/BUILD:274:1: C++ compilation of rule '@com_google_protobuf//:protoc_lib' failed (Exit 34). Note: Remote connection/protocol failed with: execution failed: com.google.devtools.build.lib.remote.Retrier$RetryException: Out of retries or status not retriable.
	at com.google.devtools.build.lib.remote.ByteStreamUploader$1.failure(ByteStreamUploader.java:245)
	at com.google.devtools.build.lib.remote.ByteStreamUploader$AsyncUpload$1.onClose(ByteStreamUploader.java:362)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:41)
	at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:339)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:443)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:63)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:525)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$600(ClientCallImpl.java:446)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:557)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:107)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: io.grpc.StatusException: INVALID_ARGUMENT: Digest mismatch c2e7bd71bf22cc3aa5b3f24693fb6b9c84ca50cb/2035 <-> 57ad52caebab9999562fde1414d8c3f65d72d993094534f2ccc4e67a5842b84d/2035
	at io.grpc.Status.asException(Status.java:534)
	at com.google.devtools.build.lib.remote.ByteStreamUploader$1.failure(ByteStreamUploader.java:238)
	... 13 more
Target //src/main/java/build/buildfarm:buildfarm-server failed to build
```

The issue is caused by hard-coded SHA1 algorithm used in [DigestUtil::compute](https://github.com/bazelbuild/bazel-buildfarm/blob/master/src/main/java/build/buildfarm/common/DigestUtil.java#L106) function. The fix is trivial and I'm working on a PR with proper tests to prevent future regressions like this.

--- a/src/main/java/build/buildfarm/common/DigestUtil.java
+++ b/src/main/java/build/buildfarm/common/DigestUtil.java
@@ -103,7 +103,7 @@ public class DigestUtil {
             public InputStream openStream() throws IOException {
               return blob.newInput();
             }
-          }.hash(Hashing.sha1()).toString(),
+          }.hash(hashFn.getHash()).toString(),
           blob.size());
     } catch(IOException ex) {
       /* impossible */

--- a/src/test/java/build/buildfarm/BuildFarmServerTest.java
+++ b/src/test/java/build/buildfarm/BuildFarmServerTest.java
@@ -19,6 +19,7 @@ import static com.google.common.truth.Truth.assertThat;
 import build.buildfarm.common.DigestUtil;
 import build.buildfarm.server.BuildFarmServer;
 import build.buildfarm.v1test.BuildFarmServerConfig;
+import build.buildfarm.v1test.InstanceHashFunction;
 import build.buildfarm.v1test.MemoryInstanceConfig;
 import com.google.devtools.remoteexecution.v1test.BatchUpdateBlobsRequest;
 import com.google.devtools.remoteexecution.v1test.BatchUpdateBlobsResponse;
@@ -70,6 +71,7 @@ public class BuildFarmServerTest {
         BuildFarmServerConfig.newBuilder().setPort(0);
     configBuilder.addInstancesBuilder()
         .setName("memory")
+        .setHashFunction(InstanceHashFunction.SHA256)
         .setMemoryInstanceConfig(memoryInstanceConfig);
 
     server = new BuildFarmServer(

--- a/src/test/java/build/buildfarm/common/BUILD
+++ b/src/test/java/build/buildfarm/common/BUILD
@@ -3,6 +3,8 @@ java_test(
     size = "small",
     srcs = ["DigestUtilTest.java"],
     deps = [
+        "//3rdparty/jvm/com/google/guava",
+        "//3rdparty/jvm/com/google/protobuf:protobuf_java",
         "//3rdparty/jvm/com/google/truth",
         "//src/main/java/build/buildfarm:common",
         "//src/main/protobuf:build_buildfarm_v1test_buildfarm_java_proto",

--- a/src/test/java/build/buildfarm/common/DigestUtilTest.java
+++ b/src/test/java/build/buildfarm/common/DigestUtilTest.java
@@ -17,6 +17,7 @@ package build.buildfarm.common;
 import static com.google.common.truth.Truth.assertThat;
 
 import com.google.devtools.remoteexecution.v1test.Digest;
+import com.google.protobuf.ByteString;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.ExpectedException;
@@ -43,5 +44,22 @@ public class DigestUtilTest {
     Digest digest = util.build(bazelMd5Hash, payloadSizeInBytes);
     assertThat(digest.getHash()).isEqualTo(bazelMd5Hash);
     assertThat(digest.getSizeBytes()).isEqualTo(payloadSizeInBytes);
+  }
+
+  @Test
+  public void computesMd5Hash() {
+    ByteString content = ByteString.copyFromUtf8("bazel");
+    DigestUtil util = new DigestUtil(DigestUtil.HashFunction.MD5);
+    Digest digest = util.compute(content);
+    assertThat(digest.getHash()).isEqualTo("24ef4c36ec66c15ef9f0c96fe27c0e0b");
+  }
+
+  @Test
+  public void computesSha256Hash() {
+    ByteString content = ByteString.copyFromUtf8("bazel");
+    DigestUtil util = new DigestUtil(DigestUtil.HashFunction.SHA256);
+    Digest digest = util.compute(content);
+    assertThat(digest.getHash())
+        .isEqualTo("aa0e09c406dd0db1a3bb250216045e81644d26c961c0e8c34e8a0354476ca6d4");
   }
 }

